### PR TITLE
All params can be arrays in OpenAPI

### LINKED ISSUE #575: Support arbitrary number of files upload
Modify OpenAPI Specification generation.
Enable isArray on file plumber type
Modify parse_multi to combine args sharing the same name

Note:
Swagger UI does not support this yet
Rapidoc UI does
Redoc does not have a try feature

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #575

Last remaining sub PR from https://github.com/rstudio/plumber/pull/578

PR task list:
- [NA] Update NEWS
- [x] Add tests
- [NA] Update documentation with `devtools::document()`

Teased out all work done by @meztez regarding all OpenAPI parameters having array support.

@meztez Does this PR look ok?

### CODE REVIEW

**Comment by meztez** [comment]
I don't think file array will work since they use the same form name and have to be combined. See file array test.

**Comment by schloerke** [comment]
@meztez Which file array test are you referring to?

**Comment by meztez** [comment]
> 
> 
> @meztez Which file array test are you referring to?

This one https://github.com/rstudio/plumber/blob/c88d4d2075c3cc37a87e3365812ab6a3b00b1465/tests/testthat/test-parse-body.R#L112

**Comment by meztez** [comment]
I'm getting
```
<simpleError: lexical error: invalid char in json text.
                                      [object File],[object File],[obj
                     (right here) ------^
>
```

I wonder if it is the first byte part... I'll investigate further

**Comment by meztez** [comment]
Something fishy is going on with the postBody : 
```R
Browse[4]> req$postBody
[1] "-----------------------------834159765195020684138062145\nContent-Disposition: form-data; name=\"f\"\n\n[object File],[object File],[object File]\n-----------------------------834159765195020684138062145--"
Browse[4]> req$postBodyRaw
  [1] 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 38 33 34 31 35 39 37 36 35 31 39 35 30 32 30 36 38 34 31 33 38 30 36 32 31 34 35 0d
 [58] 0a 43 6f 6e 74 65 6e 74 2d 44 69 73 70 6f 73 69 74 69 6f 6e 3a 20 66 6f 72 6d 2d 64 61 74 61 3b 20 6e 61 6d 65 3d 22 66 22 0d 0a 0d 0a 5b 6f 62 6a 65 63 74 20 46 69 6c 65
[115] 5d 2c 5b 6f 62 6a 65 63 74 20 46 69 6c 65 5d 2c 5b 6f 62 6a 65 63 74 20 46 69 6c 65 5d 0d 0a 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d 2d
[172] 2d 2d 2d 38 33 34 31 35 39 37 36 35 31 39 35 30 32 30 36 38 34 31 33 38 30 36 32 31 34 35 2d 2d 0d 0a
```

**Comment by meztez** [comment]
Ok, Swagger 3.25 does not support multi files upload as off yet, I had to use rapidoc.

```R
1, 0x4a, 0x57, 0x0b, 0x31, 0x33, 0x09, 0x0e, 0x0d, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82)))): formal argument "f" matched by multiple actual arguments>
```

**Comment by meztez** [comment]
```R
#
# This is a Plumber API. You can run the API by clicking
# the 'Run API' button above.
#
# Find out more about building APIs with Plumber here:
#
#    https://www.rplumber.io/
#

library(plumber)
library(rapidoc)

#* @plumber
function(pr) {
  pr$set_ui("rapidoc")
}

#* @apiTitle Plumber Example API

#* @param obj:[list]
#* @post /a
function(obj) {
  browser()
  return("ko")
}

#* Plot a histogram
#* @param f:[file]
#* @post /b
function(f) {
  browser()
  return("ko")
}
```

**Comment by meztez** [comment]
It works with swagger 3.31.1 , same formals error:

```R
formal argument "f" matched by multiple actual arguments>
```

**Comment by schloerke** [comment]
Correct. This will cause issues.

@cpsievert and I decided that there was no nice solution to the multiple values being posted. (Either make inconsistent value shapes or always make objects into lists even if it not required. Both don't sound good.  Instead, we will have the user handle multiple files with the same filename.)

Instead, we are throwing a better error message when this situation occurs: https://github.com/rstudio/plumber/pull/637/files#diff-8312ad0561ef661716b48d09478362f3R150-R153

We are encouraging the users to use `req$argsPostBody` (or `req$argsQuery`, or `req$argsPath`) which will be a named list of parameters that can have multiple named arguments.

Does this approach work / make sense?

**Comment by schloerke** [comment]
Ok. I see the issue now. 🤦 

Parsing the post body of the example gives:

```r
List of 5
 $ files:List of 1
  ..$ avatar2-small.png: raw [1:1117] 89 50 4e 47 ...
 $ files:List of 1
  ..$ text1.bin: raw 61
 $ files:List of 1
  ..$ text2.bin: raw 62
 $ files:List of 1
  ..$ text3.bin: raw 63
 $ dt   : Named list()
```

Where it should prolly have the shape of 

```r
List of 5
 $ files:List of 4
  ..$ avatar2-small.png: raw [1:1117] 89 50 4e 47 ...
  ..$ text1.bin: raw 61
  ..$ text2.bin: raw 62
  ..$ text3.bin: raw 63
 $ dt   : Named list()
```

Where we are merging on the `name` of the multi-part.

**Comment by schloerke** [comment]
Context: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

> `name`
> Is followed by a string containing the name of the HTML field in the form that the content of this subpart refers to. When dealing with multiple files in the same field (for example, the `multiple` attribute of an `<input type="file">` element), there can be several subparts with the same name.

**Comment by schloerke** [comment]
The nice part is the work done on the multiple named arguments will still be beneficial. By adopting the merging of named multipart keys, that situation should occur less often.

**Comment by schloerke** [comment]
Have updated the code to have multipart merge same `name` values.  Sorry I missed the context earlier.

Added more `combine_keys` tests

File: R/parse-query.R [hunk 1]
```diff
@@ -179,3 +151,79 @@ extractPathParams <- function(def, path){
...
+  if (length(unique_keys) == length(keys) || is.null(keys)) {
+    return(obj)
+  }
+
+  vals <- unname(obj)
+
+  extra_args <- NULL
+  if (type == "multi") {
+    # handle unnamed args by removing them from being merged and adding them back again at the end
+    browser()
```
Comment (cpsievert) [inline R/parse-query.R:172]:
```suggestion
```

File: R/parse-query.R [hunk 2]
```diff
@@ -179,3 +151,78 @@ extractPathParams <- function(def, path){
...
+      },
+    "multi" =
+      function(x) {
+        if (length(x) == 1) {
+          # return first item only
+          x[[1]]
+        } else {
+          # return list of internal named items
+          # aka... unlist the top layer only. Maintain the inner layer names
+          as.list(unlist(x, recursive = FALSE))
```
Comment (cpsievert) [inline R/parse-query.R:195]: Add in a test for this logic (maybe `x` of `list(A = list(x = 1, y = 3), A = 2, A = 5))`)

```suggestion
          as.list(unlist(unname(x), recursive = FALSE))
```

File: tests/testthat/test-combine-keys.R [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+test_that("multi part keys with same name are combined", {
+  a <- c(list(A = 1), list(A = 2), B = 3, A = 4, B = 5)
+  expect_equal(combine_keys(a, "multi"), list(A = list(1,2,4), B = list(3,5)))
+})
+
+
+test_that("multi part keys with no name are left alone", {
+  a <- c(list(A = 1), 2, list(A = 3), list(B = 4), 5)
+  expect_equal(combine_keys(a, "multi"), list(A = list(1,3), B = 4, 2, 5))
+})
```
Comment (cpsievert) [inline tests/testthat/test-combine-keys.R:17]: Let's do some more cases with lists of lists

**Pull request merged**